### ISSUE TITLE
Enhancement/handle directory changes outside oni

### ISSUE THREAD

**Issue by Akin909** [onivim/oni#1506]
If the directory is changed outside oni, whilst outside oni and theres an attempt to navigate within it in the file explorer an error is thrown that the directory does not exist. Not sure what the correct response should be but I think a potentially valid one would be to move upwards till the first valid directory is found. 

Say I create a project then change my mind or rename it Oni should hook into the focus gained command to check whether the active dir is still present in the file system if not it should *maybe* just navigate upwards the reason I suggest this is because its hopefully less back tracking than say just navigating to the home dir which would be easier but I think a bit annoying as you now have to go searching all the way down for where you were I'd think as a user you probably just want it to back out till the next reasonable location.

**Comment by Akin909** [comment]
@bryphe regarding this issue the simplest possible solution I can see would be checking if the file exists on `focusGained` in oni but this makes the assumption that this change isn't happening inside oni via a commandline command to move a file or dir.

I was thinking and wanted to run past you the idea of oni having a `fileSystem` watcher, potentially as a service to stick with the principle of one thing with a single purpose, which could then provide this functionality to areas of the app that would want to know about the status of the file system for example to file explorer the workspace etc. any thoughts?

Re. the implementation I've read around using `fs.watch` and its associates but have found that it comes with a lot of issues generally a lot of which are macOS specific alot of others more general plus its intensive on the user's CPU, does not recursively watch files and was apparently responsible for `Sublime` having issue with not watching files properly. Two alternates the first is `vscodes` solution and seems to be a well established typed solution is [chokidar](https://github.com/paulmillr/chokidar) the other is [`@atom/watcher`](https://github.com/atom/watcher) which is atom's home grown apparently highly performant bleeding edge alternative (side note, tried using this earlier and lead to a node_modules implosion...), I'm happy to proceed in whatever direction would serve oni best.

**Comment by bryphe** [comment]
Thanks for thinking about this and putting together this proposal, @Akin909 .

> I was thinking and wanted to run past you the idea of oni having a fileSystem watcher, potentially as a service to stick with the principle of one thing with a single purpose, which could then provide this functionality to areas of the app that would want to know about the status of the file system for example to file explorer the workspace etc. any thoughts?

I think this sounds reasonable. I know I've hit this where I have a folder expanded in the explorer, I save a new file, and I don't see it in the explorer until I collapse/expand the folder. Would be great to have a story for that!

> Re. the implementation I've read around using fs.watch and its associates but have found that it comes with a lot of issues generally a lot of which are macOS specific alot of others more general plus its intensive on the user's CPU, does not recursively watch files and was apparently responsible for Sublime having issues with not watching files properly. 

Makes sense... we actually have issues like that today with our configuration files. `chokidar` seems like a robust, proven solution. That sounds like a good one to integrate!

The next piece to think about is what the interface of this `Service` would look like and how it would interact with the other components. What would its API be? Where would we hook into it with the explorer? On the explorer side, it's pretty simple to dispatch a new `REFRESH` event in response if we know something updated.

**Comment by Akin909** [comment]
@bryphe re the API I was thinking that what `chokidar` provides is essentially an event emitter, which informs re status updates to the file system and that a service based on that could essentially be an Oni wrapper around the emissions aka the api would potentially look like 
```js
	Oni.editors.activeEditor.fileSystem(nameTBD).fileChanged.subscribe(
		fileChange => oniActionBasedOn(fileChanges)
	)

	Oni.editors.activeEditor.fileSystem(nameTBD).fileDeleted.subscribe(
		fileDeleted => oniActionBasedOn(fileDeleted)
	)

	Oni.editors.activeEditor.fileSystem(nameTBD).folderChanged.subscribe(
		folderChanged => oniActionBasedOn(folderChanged)
	)
	...
```
so each of the different events could be mapped to a subscription so a consumer can subscribe to the type of events it is interested in, potentially this could be generalised in some way but ideally you wouldn't have to listen to events that didn't relate to what you wanted to know so it might make subscriptions tedious if you want to listen to multiple types of events.

possibly another potential method would be

```
	Oni.editors.activeEditor.fileSystem(nameTBD).allFileEvents.subscribe(
		objectOfDifferentTypesOfChangesAndEachChange => oniActionBas
edOn(fileChanges)
	)
```

Thats as far as I've gotten re. how this would work I'm very open to suggestions as I don't have a strong preference or path in mind

**Comment by keforbes** [comment]
Addressed in #2116, closing.
